Clarifying tag key parameters (#35343)
DataDog/documentation: content/en/getting_started/tagging/_index.md

Tags are a way of adding dimensions to Datadog telemetries so they can be filtered, aggregated, and compared in Datadog visualizations. [Using tags][1] enables you to observe aggregate performance across several hosts and (optionally) narrow the set further based on specific elements. In summary, tagging is a method to observe aggregate datapoints.

> A tag can be formatted as `<key>:<value>` or `<value>`. It is recommended to use the `<key>:<value>` format, as it is often semantically clearer and allows for richer querying capabilities (for example, grouping by key). When using a `<key>:<value>` pair:

Reviewer: ```suggestion
A tag can be formatted as `<key>:<value>` or `<value>`. Datadog recommends using the `<key>:<value>` format, as it is often semantically clearer and allows for richer querying capabilities (for example, grouping by key). When using a `<key>:<value>` pair:
```

## Define tags

Tag strings (that is, the entire content of `<key>:<value>` or `<value>`) must meet the following requirements:

- Tag strings must **start with a letter** (this applies regardless of whether the tag uses the format `<key>:<value>` or `<value>`). After the leading letter, the tag string may contain the characters listed below:

>     - Letters (all Unicode letters are supported---for example, a, ó, 気, 녕, ك, and ดี)

Reviewer: ```suggestion
    - Letters (all Unicode letters are supported—for example, a, ó, 気, 녕, ك, and ดี)
```

Reviewer: Docs typically use the Unicode em-dash character — rather than ---. Example from a sibling file (content/en/getting_started/tagging/using_tags.md): 
`"narrow down the scope even further—for example…"`

    - Numbers

>     - Underscores (leading/trailing underscores are stripped, and contiguous underscores are collapsed into one)

Reviewer: ```suggestion
    - Underscores (leading and trailing underscores are removed, and contiguous underscores are collapsed into one)
```

    - Minuses
    - Colons
    - Periods
    - Forward slashes

>     - (For tags on logs [ingested via HTTP][28] only) at signs (@)

Reviewer: ```suggestion
    - At signs (`@`)—only for tags on logs [ingested through HTTP][28]
```

Reviewer: Moving the parenthetical to the end makes this a bit easier to scan

Author: thanks! I can see how switching the order makes this list item easier to scan, but I wonder if it carries a cost—like people skimming past the parenthetical

in general I try to put conditional clauses before the instruction or information they modify (see this concept further explained on Google's [Sentence Structure](https://developers.google.com/style/sentence-structure) page)

Author: I applied your suggestions to switch `via` to `through` and to code format the `@`, but left the parenthetical at the beginning of the text. let me know whatcha think @buraizu !

Reviewer: Thanks for the link — that page is a useful reference, but I'd push back a bit on whether it applies here.
                                                                                                                                                                  
The Google guidance is explicitly about instructions: "If you want to tell the reader to do something, try to mention the circumstance...before you provide the instruction" (their example is "Click Delete if you want to delete the entire document" → "To delete the entire document, click Delete"). The bullet at line 68 isn't an instruction, though — it's an item in a noun list enumerating allowed characters. The circumstance-before instruction rule is about procedural sentences, not list items.                                                                                                                                      
                  
The trade-off I see: every other bullet in this list leads with the character class as the subject ("Letters", "Numbers", "Underscores", and so on). The @  bullet is the only one that inverts that, so a reader scanning to answer "is @ allowed?" has to read the parenthetical before discovering which bullet they're on. The skim-past risk applies in either ordering — readers can skim past a trailing qualifier or a leading parenthetical equally — but the parallelism cost is paid on every scan.

If the concern is keeping the conditional prominent without breaking parallelism, one option is to promote it out of the bullet:                                
   

> - At signs (@)                                                                                                                                                
>       
> **Note**: At signs are allowed only in tags on logs [ingested through HTTP][28].

Up to you.

Author: thanks! the Google page does focus on instructions, but the third example shows how the principle can be applied to non-instructional information:

- bad: Using custom domains might add noticeable latency to responses if your app is located in one of the following regions:
- good: If your app is located in one of the following regions, using custom domains might add noticeable latency to responses:

> The skim-past risk applies in either ordering

this is true, but I think the negative impact of using an `@` sign in a non-HTTP-ingested log tag is greater than someone missing that they actually _can_ use `@`s in a HTTP-ingested-log tag. in the former, they'll generate an error in Datadog, and have to go back and rename the tag

as such, I'm going to leave it as is. thank you for this discussion, @buraizu—I enjoy digging into these things, especially with someone with a slightly different perspective!

>     All other characters (including, but not limited to, commas, emoji, back slashes, and spaces) are converted to underscores.

Reviewer: ```suggestion
    All other characters (including commas, emoji, backslashes, and spaces) are converted to underscores.
```

Reviewer: My perception is that `including` already implies `but not limited to`, but open to feedback if you have a different perspective

Author: I use `but not limited to` when I want to stress that the list is non-exhaustive, but I don't think it's super important to do so in this case. applying your suggestion, thanks!

    
    **Notes**:
    - A tag that starts with a digit may be accepted in some contexts, such as `env` tags set at the Agent level. However, tags that don't follow standard naming rules may not work consistently across all Datadog products and can increase tag cardinality. Start tags with a letter unless a specific product explicitly supports otherwise.
    - The `DD_TAGS` environment variable uses whitespace as a separator between tags. Whitespace in `DD_TAGS` values is **not** converted to underscores. For example, `DD_TAGS="test:this is a test"` produces four separate tags: `test:this`, `is`, `a`, and `test`. To set a tag value that contains spaces, use a YAML configuration file or integration annotations, where whitespace is converted to underscores.

> - Tags can be **up to 200 characters** long. If the tag has the format `<key>:<value>`, the `<key>`, `:`, and `<value>` all count toward the character limit.

Reviewer: ```suggestion
- Tags can be **up to 200 characters** long. If the tag has the format `<key>:<value>`, the key, `:`, and value all count toward the character limit.
```

Reviewer: Minor update to untangle conceptually different uses of "key" and "value". Alternatively, since the key/value concept is already defined earlier in the page, this could be shortened to:
`Tags can be up to 200 characters long, including the key, the :, and the value.`

Author: nice catch! I took your suggestion

> - [Span tags][26] and metric tags are normalized to lowercase. Because of this, using `CamelCase` is not recommended. Authentication (crawler) based integrations also convert camel case tags to underscores before lowercasing, for example `TestTag` --> `test_tag`.

Reviewer: ```suggestion
- [Span tags][26] and metric tags are normalized to lowercase, so avoid camelCase. Authentication (crawler) based integrations also convert camel-case tags to underscores before lowercasing. For example, `testTag` becomes `test_tag`.
```

Reviewer: These examples aren't actually camel case, and it's worth being precise here because these are identifier-convention terms with established technical meanings.

Additionally, `-->` is an HTML comment sentinel and renders inconsistently, and using prose might be more understandable for screen readers.

Author: thanks! I made an update to use "camel case," since it seems like that's what we've used elsewhere in the docs. (I like the idea of styling it autologically, but didn't feel it was imperative)

>     - Separate from tags, [span attributes][27] and log attributes are case-sensitive, and are not normalized.

Reviewer: ```suggestion
    - Unlike tags, [span attributes][27] and log attributes are case-sensitive, and are not normalized.
```

- When using the format `<key>:<value>`, the key always precedes the first colon of the global tag definition. For example:
    
    | Tag                | Key           | Value          |
    | ------------------ | ------------- | -------------- |
    | `env:staging:east` | `env`         | `staging:east` |
    | `env_staging:east` | `env_staging` | `east`         |

> - Tags should not originate from unbounded sources, such as epoch timestamps, user IDs, or request IDs. Doing so may infinitely [increase the number of metrics][2] for your organization and impact your billing.

Reviewer: ```suggestion
- Tags should not originate from unbounded sources, such as epoch timestamps, user IDs, or request IDs. Doing so can cause unbounded growth in the number of metrics.
```

Reviewer: Optional suggestion since this is not actually part of your changes

Author: thanks, I'm down to apply this suggestion, but also just curious—did you remove the clause about billing being impacted because that's implied?

Author: I took most of your suggestion, but left in the link (but modified it to only apply to "metrics")

Below are Datadog's tagging requirements:

- Tags must **start with a letter** (this applies regardless of whether the tag uses the format `<key>:<value>` or `<value>`). After the leading letter, the tag string may contain the characters listed below:

    - Letters (all Unicode letters are supported---for example, a, ó, 気, 녕, ك, and ดี)
    - Numbers
    - Underscores (leading/trailing underscores are stripped, and contiguous underscores are collapsed into one)
    - Minuses
    - Colons
    - Periods

>     - Forward slashes

Author: hi @nhinsch — thanks for [volunteering](https://dd.slack.com/archives/C03A4P0207Q/p1775164129480529?thread_ts=1774618563.282709&cid=C03A4P0207Q) to jump in here in Ray's stead! I'm hoping to move this change along, so please provide any additional feedback by Monday, 4/13. as of now, I'm planning to leave @ out of the allowlist, per my Claude-assisted research. as I mentioned [here](https://dd.slack.com/archives/C03A4P0207Q/p1775224159711669?thread_ts=1774618563.282709&cid=C03A4P0207Q), if there are additional important, log-specific distinctions, feel free to share them! thank you!

>     - Forward slashes

Reviewer 2: To circle back on this, it turns out that in logs, `@` is allowed by our backend, but is removed by the Datadog agent. Nonetheless I will approve the PR and let you decide how you want to proceed with this docs change 👍

>     - Forward slashes

Author: thanks! as I said [here](https://dd.slack.com/archives/C07FZ0UG4CC/p1776866784092949?thread_ts=1776146071.343129&cid=C07FZ0UG4CC), I think it makes sense to specify in the docs that @s in tags for logs ingested via HTTP are allowed/not normalized to _s. I just pushed a commit that makes that change

with that, I think we're ready to publish this!

    - Minuses
    - Colons
    - Periods
    - Forward slashes
    - (Only for tags on logs [ingested through HTTP][28]) at signs (`@`)

    All other characters (including commas, emoji, backslashes, and spaces) are converted to underscores.
    
    **Notes**:
    - A tag that starts with a digit may be accepted in some contexts, such as `env` tags set at the Agent level. However, tags that don't follow standard naming rules may not work consistently across all Datadog products and can increase tag cardinality. Start tags with a letter unless a specific product explicitly supports otherwise.
    - The `DD_TAGS` environment variable uses whitespace as a separator between tags. Whitespace in `DD_TAGS` values is **not** converted to underscores. For example, `DD_TAGS="test:this is a test"` produces four separate tags: `test:this`, `is`, `a`, and `test`. To set a tag value that contains spaces, use a YAML configuration file or integration annotations, where whitespace is converted to underscores.

- Tags can be **up to 200 characters** long. If the tag has the format `<key>:<value>`, the key, `:`, and value all count toward the character limit.

> - [Span tags][26] and metric tags are normalized to lowercase, so avoid using camel case. Authentication (crawler) based integrations also convert camel case tags to underscores before lowercasing. For example `testTag` becomes `test_tag`.
>     - Unlike tags, [span attributes][27] and log attributes are case-sensitive, and are not normalized.

Reviewer: ```suggestion
- [Span tags][26] and metric tags are normalized to lowercase, so avoid using camel case. Authentication (crawler) based integrations also convert camel case tags to underscores before lowercasing. For example, `testTag` becomes `test_tag`.
    - Unlike tags, [span attributes][27] and log attributes are case-sensitive and are not normalized.
```
